README.md
# whooohoo jackpot

[![build status](https://secure.travis-ci.org/3rd-Eden/jackpot.png?branch=master)](http://travis-ci.org/3rd-Eden/jackpot)

Jackpot is a fault tolerant connection pool for Node.js, it automatically cleans
up after it self and detects broken connections. It does not need to be
released, as it will allocate connections based on their readyState / write
abilities.

## API

```js
var ConnectionPool = require('jackpot');

// first argument: size of the connection pool
// second argument: optional connection factory
var pool = new ConnectionPool(100);

// you can optionally boost the amount of retries after you have
// constructed a new instance
pool.retries = 5 // allow 5 failures for the #pull method

// every connection pool requires a factory which is used to generate / setup
// the initial net.Connection
//
// it should return a new net.Connection instance..
pool.factory(function () {
  return net.connect(port, host)
});

// now that the pool is setup we can allocate a connection, the allocate
// requires a callback as it can be async..
pool.allocate(function (err, connection) {
  // error: when we failed to get a connection
  // connection: the allocated net.connection if there isn't an error
});

// in addition to the #allocate method, there is also the pull method, which is
// fault tolerant wrapper around the allocate function as it can give back an
// error if the pool is full, so the pull will then retry to get a new
connection.pull(function (err, connection) {
  // same arguments as the #allocate method
});

// call pool.free if you want to free connections from the pool, the arugment
// you supply is the amount of connections you want to keep
pool.free(10); // keep only 10 healthy connections kill the rest.

// kill the whole connection pool:
pool.end();
```

For more API information, fork this repo and add more.. or look at the test
file.

## LICENSE (MIT)

Copyright (c) 2012 Observe.it (http://observe.it) <opensource@observe.it>

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions: 

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
#### v0.0.5
- Adding missing `timeout` listener. See #9

#### v0.0.4
- Added support for configurable options.
- Fixed a bug where broken connections were briefly added to the pool

#### v0.0.3
- Changed the way how we inherit from the `EventEmitter` prototype to combat
  potential memory leaks. See #5

#### v0.0.2
- Fixed a socket leak that was caused by incorrectly splicing an array. It
  removed the first item from the pool instead of item that was found using
  indexOf. See #2

#### v0.0.1
- Small fix for the isAvailable method, returns a larger availability. See #1

#### v0.0.0
- Initial release


test/jackpot.test.js
/*globals expect, TESTNUMBER, net */
describe('jackpot', function () {
  'use strict';

  var ConnectionPool = require('../')
    , port = TESTNUMBER
    , host = 'localhost'
    , server;

  before(function before(done) {
    server = net.createServer();
    server.listen(port, host, done);
  });

  it('should exported as a function', function () {
    expect(ConnectionPool).to.be.a('function');
  });

  it('should be an instance of EventEmitter', function () {
    expect(new ConnectionPool).to.be.instanceof(process.EventEmitter);
  });

  describe('initialize', function () {
    it('should update the limit if its supplied', function () {
      var pool = new ConnectionPool(100);

      expect(pool.limit).to.equal(100);
    });

    it('should also update the generator if its supplied', function () {
      function test() {}
      var pool = new ConnectionPool(100, test);

      expect(pool.generator).to.equal(test);
    });

    it('applies the options', function () {
      function test() {}
      var pool = new ConnectionPool(100, test, {
        retries: 10,
        factor: 10,
        randomize: false
      });

      expect(pool.generator).to.equal(test);
      expect(pool.retries).to.equal(10);
      expect(pool.factor).to.equal(10);
      expect(pool.randomize).to.equal(false);
    });

    it('allows options to be second argument', function () {
      var pool = new ConnectionPool(10, {
        retries: 10
      });

      expect(pool.retries).to.equal(10);
    });

    it('parses human readable strings', function () {
      var pool = new ConnectionPool(10, {
        min: '10 seconds',
        max: 10000
      });

      expect(pool.minTimeout).to.equal(10 * 1000);
      expect(pool.maxTimeout).to.equal(10000);
    });
  });

  describe('#factory', function () {
    it('should throw an error if the supplied factory is not a function', function () {
      var pool = new ConnectionPool();

      expect(pool.factory).to.throw(Error);
    });

    it('should not throw an error if a function is supplied', function () {
      var pool = new ConnectionPool();

      function test() {}
      expect(pool.factory.bind(pool, test)).to.not.throw(Error);
    });
  });

  describe('#allocate', function () {
    it('should give an error when no #factory is specified', function (done) {
      var pool = new ConnectionPool();

      pool.allocate(function allocate(err, conn) {
        expect(err).to.be.an.instanceof(Error);
        expect(err.message).to.contain('#factory');

        done();
      });
    });

    it('should emit an error when we cannot establish a connection', function (done) {
      var pool = new ConnectionPool()
        , differentport = TESTNUMBER;

      pool.once('error', function error(err) {
        expect(err).to.be.an.instanceof(Error);
      });

      pool.factory(function factory() {
        return net.connect(differentport, host);
      });

      // make sure the port is different
      expect(differentport).to.not.eql(port);

      pool.allocate(function allocate(err, connection) {
        expect(err).to.be.an.instanceof(Error);

        done();
      });
    });

    it('should NOT emit an error when we can establish a connection', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.allocate(function allocate(err, connection) {
        expect(err).to.not.be.an.instanceof(Error);
        expect(connection).to.be.an.instanceof(net.Socket);

        connection.end();
        done();
      });
    });

    it('should an error occure, then remove it from the pool', function (done) {
      var pool = new ConnectionPool()
        , differentport = TESTNUMBER;

      pool.once('error', function error(err) {
        expect(pool.pool).to.have.length(0);

        done();
      });

      pool.factory(function factory() {
        return net.connect(differentport, host);
      });

      // make sure the port is different
      expect(differentport).to.not.eql(port);

      pool.allocate(function allocate(err, connection) {});
    });

    it('should increase poolsize when a connection is allocated', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.allocate(function allocate(err, connection) {
        expect(connection).to.be.an.instanceof(net.Socket);
        expect(pool.pool).to.have.length(1);

        // make sure it also decreases when the connection is closed
        connection.on('end', function end() {
          expect(pool.pool).to.have.length(0);

          done();
        });
        connection.end();
      });
    });

    it('should increase metrics when allocating a connection', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.allocate(function allocate(err, connection) {
        expect(pool.metrics.allocations).to.eql(1);

        // make sure it also decreases when the connection is closed
        connection.on('end', function end() {
          expect(pool.metrics.allocations).to.eql(1);
          expect(pool.metrics.releases).to.eql(1);

          done();
        });

        connection.end();
      });
    });

    it('should handle burst allocations', function (done) {
      var pool = new ConnectionPool()
        , count = 0
        , allocations = 50;

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      /**
       * Small helper function.
       *
       * @param {error} err
       * @api private
       */

      function allocate(err, conn) {
        if (++count !== allocations) return;

        // free all the things
        pool.end(true);
        expect(pool.pending).to.eql(0);
        expect(pool.pool).to.have.length(0);
      }

      for (var i = 0; i < allocations; i++) {
        pool.allocate(allocate);
      }

      pool.once('free', function (saved) {
        expect(saved).to.eql(0);
      });

      pool.once('end', done);
    });
  });

  describe('#pull', function () {
    it('should give an error when no #factory is specified', function (done) {
      var pool = new ConnectionPool(10, { retries: 1 });

      pool.pull(function allocate(err, conn) {
        expect(err).to.be.an.instanceof(Error);
        expect(err.message).to.contain('#factory');

        done();
      });
    });

    it('should honor the retry configuration', function (done) {
      this.timeout(500);

      var pool = new ConnectionPool(10, {
        min: 10,
        retries: 1
      });

      pool.pull(function allocate(err, conn) {
        expect(err).to.be.an.instanceof(Error);
        expect(err.message).to.contain('#factory');

        done();
      });
    });

    it('should NOT emit an error when we can establish a connection', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.pull(function allocate(err, connection) {
        expect(err).to.not.be.an.instanceof(Error);
        expect(connection).to.be.an.instanceof(net.Socket);

        connection.end();
        done();
      });
    });

    it('should an error occure, then remove it from the pool', function (done) {
      var pool = new ConnectionPool()
        , differentport = TESTNUMBER;

      pool.once('error', function error(err) {
        expect(pool.pool).to.have.length(0);

        done();
      });

      pool.factory(function factory() {
        return net.connect(differentport, host);
      });

      // make sure the port is different
      expect(differentport).to.not.eql(port);

      pool.pull(function allocate(err, connection) {});
    });

    it('should increase poolsize when a connection is allocated', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.pull(function allocate(err, connection) {
        expect(connection).to.be.an.instanceof(net.Socket);
        expect(pool.pool).to.have.length(1);

        // make sure it also decreases when the connection is closed
        connection.on('end', function end() {
          expect(pool.pool).to.have.length(0);

          done();
        });
        connection.end();
      });
    });

    it('should increase metrics when allocating a connection', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.pull(function allocate(err, connection) {
        expect(pool.metrics.allocations).to.eql(1);

        // make sure it also decreases when the connection is closed
        connection.on('end', function end() {
          expect(pool.metrics.allocations).to.eql(1);
          expect(pool.metrics.releases).to.eql(1);

          done();
        });

        connection.end();
      });
    });
  });

  describe('#forEach', function () {
    it('should iterate over the items in the pool');
    it('should set the given context');
  });

  describe('#free', function () {
    it('should kill all connections when 0 is given', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.allocate(function allocate(err, connection) {
        expect(connection).to.be.an.instanceof(net.Socket);
        expect(pool.pool).to.have.length(1);

        pool.once('free', function free(saved, size) {
          expect(pool.pool).to.have.length(0);
          expect(size).to.eql(pool.pool.length);
          expect(saved).to.eql(0);

          done();
        });

        pool.free(0);
      });
    });

    it('should keep 1 connection', function (done) {
      var pool = new ConnectionPool();

      pool.factory(function factory() {
        return net.connect(port, host);
      });

      pool.allocate(function allocate(err, connection) {
        expect(connection).to.be.an.instanceof(net.Socket);
        expect(pool.pool).to.have.length(1);

        pool.once('free', function free(saved, size) {
          expect(pool.pool).to.have.length(1);
          expect(size).to.eql(pool.pool.length);
          expect(saved).to.eql(1);

          connection.end();
          done();
        });

        pool.free(1);
      });
    });
  });

  after(function after(done) {
    // in 0.7, we can supply the server with the done callback
    server.once('close', done);
    server.close();
  });
});


lib/jackpot.js
'use strict';

var EventEmitter = require('events').EventEmitter
  , retry = require('retry');

/**
 * A net.Stream connection pool.
 *
 * @constructor
 * @param {Number} limit size of the connection pool
 * @param {Function} builder stream factory
 * @api public
 */

function Manager(limit, builder) {
  this.limit = +limit || 20; // defaults to 20 connections max
  this.pool = [];
  this.pending = 0;
  this.generator = null;
  this.retries = 5;

  // some stats that can be used for metrics
  this.metrics = {
      allocations: 0
    , releases: 0
  };

  if (builder) this.factory(builder);
  EventEmitter.call(this);
}

Manager.prototype = new EventEmitter();
Manager.prototype.constructor = Manager;

/**
 * Add a stream generator so we can generate streams for the pool.
 *
 * @param {Function} builder
 * @api public
 */

Manager.prototype.factory = function factory(builder) {
  if (typeof builder !== 'function') {
    throw new Error('The #factory requires a function');
  }

  this.generator = builder;
};

/**
 * Start listening to events that could influence the state of the connection.
 *
 * @param {net.Connection} net
 * @api private
 */

Manager.prototype.listen = function listen(net) {
  if (!net) return;

  var self = this;

  /**
   * Simple helper function that allows us to automatically remove the
   * connection from the pool when we are unable to connect using it.
   *
   * @param {Error} err optional error
   * @api private
   */

  function regenerate(err) {
    net.destroySoon();

    self.remove(net);
    net.removeListener('error', regenerate);
    net.removeListener('end', regenerate);

    if (err) self.emit('error', err);
  }

  // listen for events that would mess up the connection
  net.on('error', regenerate)
     .on('end', regenerate);
};

/**
 * A fault tolerant connection allocation wrapper.
 *
 * @param {Function} fn
 * @api private
 */

Manager.prototype.pull = function pull(fn) {
  var operation = retry.operation({
          retries: this.retries
        , factor: 3
        , minTimeout: 1 * 1000
        , maxTimeout: 60 * 1000
        , randomize: true
      })
    , self = this;

  /**
   * Small wrapper around pulling a connection
   *
   * @param {Error} err
   * @param {Socket} connection
   * @api private
   */

  function allocate(err, connection) {
    if (operation.retry(err)) return;

    fn(err, connection);
  }

  operation.attempt(function attempt() {
    self.allocate(allocate);
  });
};

/**
 * Allocate a new connection from the connection pool, this can be done async
 * that's why we use a error first callback pattern.
 *
 * @param {Function} fn
 * @api public
 */

Manager.prototype.allocate = function allocate(fn) {
  if (!this.generator) return fn(new Error('Specify a stream #factory'));

  /**
   * Small helper function that allows us to correctly call the callback with
   * the correct arguments when we generate a new connection as the connection
   * should be emitting 'connect' befor we can use it. But it can also emit
   * error if it fails to connect.
   *
   * @param {Error} err
   * @api private
   */

  function either(err) {
    this.removeListener('error', either);
    this.removeListener('connect', either);

    // add to the pool
    self.pool.push(this);
    self.pending--;

    fn(err, this);
  }

  var probabilities = []
    , self = this
    , total, i, probability, connection;

  i = total = this.pool.length;

  // increase the allocation metric
  this.metrics.allocations++;

  // check the current pool if we already have a few connections available, so
  // we don't have to generate a new connection
  while (i--) {
    connection = this.pool[i];
    probability = this.isAvailable(connection);

    // we are sure this connection works
    if (probability === 100) return fn(undefined, connection);

    // no accurate match, add it to the queue as we can get the most likely
    // available connection
    probabilities.push({
        probability: probability
      , connection: connection
    });
  }

  // we didn't find a confident match, see if we are allowed to generate a fresh
  // connection
  if ((this.pool.length + this.pending) < this.limit) {
    // determin if the function expects a callback or not, this can be done by
    // checking the length of the given function, as the amount of args accepted
    // equals the length..
    if (this.generator.length === 0) {
      connection = this.generator();

      if (connection) {
        this.pending++;
        this.listen(connection);
        return connection.on('error', either).on('connect', either);
      }
    } else {
      return this.generator(function generate(err, connection) {
        if (err) return fn(err);
        if (!connection) return fn(new Error('The #factory failed to generate a stream'));

        self.pending++;
        self.listen(connection);
        return connection.on('error', either).on('connect', either);
      });
    }
  }

  // o, dear, we got issues.. we didn't find a valid connection and we cannot
  // create more.. so we are going to check if we might have semi valid
  // connection by sorting the probabilities array and see if it has
  // a probability above 60
  probability = probabilities.sort(function sort(a, b) {
    return a.probability - b.probability;
  }).pop();

  if (probability && probability.probability >= 60) {
    return fn(undefined, probability.connection);
  }

  // well, that didn't work out, so assume failure
  fn(new Error('The connection pool is full'));
};

/**
 * Check if a connection is available for writing.
 *
 * @param {net.Connection} net
 * @param {Boolean} ignore ignore closed or dead connections
 * @returns {Number} probability that his connection is available or will be
 * @api private
 */

Manager.prototype.isAvailable = function isAvailable(net, ignore) {
  var readyState = net.readyState
    , writable = readyState === 'open' || readyState === 'writeOnly'
    , writePending = net._pendingWriteReqs || 0
    , writeQueue = net._writeQueue || []
    , writes = writeQueue.length || writePending;

  // if the stream is writable and we don't have anything pending we are 100%
  // sure that this stream is available for writing
  if (writable && writes === 0) return 100;

  // the connection is already closed or has been destroyed, why on earth are we
  // getting it then, remove it from the pool and return 0
  if (readyState === 'closed' || net.destroyed) {
    this.remove(net);
    return 0;
  }

  // if the stream isn't writable we aren't that sure..
  if (!writable) return 0;

  // the connection is still opening, so we can write to it in the future
  if (readyState === 'opening') return 70;

  // we have some writes, so we are going to substract that amount from our 100
  if (writes < 100) return 100 - writes;

  // we didn't find any reliable states of the stream, so we are going to
  // assume something random, because we have no clue, so generate a random
  // number between 0 - 70
  return Math.floor(Math.random() * 70);
};

/**
 * Release the connection from the connection pool.
 *
 * @param {Stream} net
 * @param {Boolean} hard destroySoon or destroy
 * @returns {Boolean} was the removal successful
 * @api private
 */

Manager.prototype.release = function release(net, hard) {
  var index = this.pool.indexOf(net);

  // no match
  if (index === -1) return false;

  // check if the stream is still open
  if (net) {
    if (!hard) net.destroySoon();
    else net.destroy();

    // remove it from the pool
    this.pool.splice(net, 1);

    // increase the releases metric
    this.metrics.releases++;
  }

  return true;
};

// alias remove to release
Manager.prototype.remove = Manager.prototype.release;

/**
 * Free dead connections from the pool.
 *
 * @param {Number} keep the amount of connection to keep open
 * @param {Boolean} hard destroy all connections instead of destroySoon
 * @api public
 */

Manager.prototype.free = function free(keep, hard) {
  // default to 0 if no arguments are supplied
  keep = +keep || 0;

  // create a back-up of the pool as we will be removing items from the array
  // and this could cause memory / socket leaks as we are unable to close some
  // connections in the array as the index has moved.
  var pool = this.pool.slice(0)
    , saved = 0;

  for (var i = 0, length = pool.length; i < length; i++) {
    var connection = pool[i]
      , probability = this.isAvailable(connection);

    // this is still a healthy connection, so try we probably just want to keep it
    if (keep && saved < keep && probability === 100) {
      saved++;
      continue;
    }

    this.release(connection, hard);
  }

  // clear the back-up
  pool.length = 0;

  // see how much connections are still available
  this.emit('free', saved, this.pool.length);
};

/**
 * Close the connection pool.
 *
 * @param {Boolean} hard destroy all connections
 * @api public
 */

Manager.prototype.end = function end(hard) {
  this.free(0, hard);

  return this.emit('end');
};

module.exports = Manager;


index.js
'use strict';

var retry = require('retry')
  , ms = require('./ms');

/**
 * A net.Stream connection pool.
 *
 * @constructor
 * @param {Number} limit Size of the connection pool.
 * @param {Function} builder Stream factory.
 * @param {Object} options Configuration options.
 * @api public
 */
function Manager(limit, builder, options) {
  options = options || {};

  if ('object' === typeof builder) {
    options = builder;
    builder = null;
  }

  this.limit = +limit || 20;                        // Defaults to 20 connections max.
  this.pool = [];                                   // Contains the active connections.
  this.pending = 0;                                 // Connecting connections.
  this.generator = null;                            // Stream generator.
  this.retries = options.retries || 5;              // Max retries for each #pull.
  this.factor = options.factor || 3;                // Exponential backoff factor.
  this.minTimeout = ms(options.min || '1 seconds'); // Minimum delay.
  this.maxTimeout = ms(options.max || '60 seconds');// Maximum delay.
  this.randomize = 'randomize' in options           // Randomize the retry timeout.
    ? options.randomize
    : true;

  // Some stats that can be used for metrics.
  this.metrics = {
      allocations: 0
    , releases: 0
  };

  if (builder) this.factory(builder);
}

Manager.prototype.__proto__ = require('events').EventEmitter.prototype;

/**
 * Add a stream generator so we can generate streams for the pool.
 *
 * @param {Function} builder Function that pre-configures a stream.
 * @api public
 */
Manager.prototype.factory = function factory(builder) {
  if (typeof builder !== 'function') {
    throw new Error('The #factory requires a function.');
  }

  this.generator = builder;
  return this;
};

/**
 * Start listening to events that could influence the state of the connection.
 *
 * @param {net.Connection} net The connection.
 * @api private
 */
Manager.prototype.listen = function listen(net) {
  if (!net) return this;

  var self = this;

  /**
   * Simple helper function that allows us to automatically remove the
   * connection from the pool when we are unable to connect using it.
   *
   * @param {Error} err Optional error.
   * @api private
   */
  function regenerate(err) {
    net.destroySoon();

    self.remove(net);

    net.removeListener('timeout', regenerate);
    net.removeListener('error', regenerate);
    net.removeListener('end', regenerate);

    if (err) self.emit('error', err);
  }

  // Listen for events that would mess up the connection.
  net.once('timeout', regenerate)
     .once('error', regenerate)
     .once('end', regenerate);

  return this;
};

/**
 * A fault tolerant connection allocation wrapper.
 *
 * @param {Function} fn Callback that receives a working connection.
 * @api public
 */
Manager.prototype.pull = function pull(fn) {
  var operation = retry.operation({
          retries: this.retries
        , factor: this.factor
        , minTimeout: this.minTimeout
        , maxTimeout: this.maxTimeout
        , randomize: this.randomize
      })
    , self = this;

  /**
   * Small wrapper around pulling a connection.
   *
   * @param {Error} err Optional error argument.
   * @param {Socket} connection A working connection.
   * @api private
   */
  function allocate(err, connection) {
    if (operation.retry(err)) return;

    fn(err, connection);
  }

  operation.attempt(function attempt() {
    self.allocate(allocate);
  });

  return this;
};

/**
 * Allocate a new connection from the connection pool, this can be done async
 * that's why we use a error first callback pattern.
 *
 * @param {Function} fn
 * @api public
 */
Manager.prototype.allocate = function allocate(fn) {
  if (!this.generator) {
    fn(new Error('Specify a stream #factory'));
    return this;
  }

  /**
   * Two helper functions that allows us to correctly call the callback with
   * the correct arguments when we generate a new connection as the connection
   * should be emitting 'connect' befor we can use it. But it can also emit
   * error if it fails to connect, or times in so doing.
   *
   * @param {Error} err Optional error argument.
   * @api private
   */
  function either(err) {
    this.removeListener('error', either);
    this.removeListener('connect', either);
    this.removeListener('timeout', timeout);

    // Add to the pool
    if (!err) self.pool.push(this);
    self.pending--;

    fn(err, this);
  }

  function timeout() {
    this.removeListener('timeout', timeout);
    self.pending--;
    fn(new Error('Timed out while trying to establish connection'), this);
  }

  var probabilities = []
    , self = this
    , total, i, probability, connection;

  i = total = this.pool.length;

  // increase the allocation metric
  this.metrics.allocations++;

  // check the current pool if we already have a few connections available, so
  // we don't have to generate a new connection
  while (i--) {
    connection = this.pool[i];
    probability = this.isAvailable(connection);

    // we are sure this connection works
    if (probability === 100) {
      fn(undefined, connection);
      return this;
    }

    // no accurate match, add it to the queue as we can get the most likely
    // available connection
    probabilities.push({
        probability: probability
      , connection: connection
    });
  }

  // we didn't find a confident match, see if we are allowed to generate a fresh
  // connection
  if ((this.pool.length + this.pending) < this.limit) {
    // determin if the function expects a callback or not, this can be done by
    // checking the length of the given function, as the amount of args accepted
    // equals the length..
    if (this.generator.length === 0) {
      connection = this.generator();

      if (connection) {
        this.pending++;
        this.listen(connection);
        connection.on('error', either)
          .on('connect', either)
          .on('timeout', timeout);

        return this;
      }
    } else {
      return this.generator(function generate(err, connection) {
        if (err) return fn(err);
        if (!connection) return fn(new Error('The #factory failed to generate a stream'));

        self.pending++;
        self.listen(connection);
        return connection.on('error', either)
          .on('connect', either)
          .on('timeout', timeout);
      });
    }
  }

  // o, dear, we got issues.. we didn't find a valid connection and we cannot
  // create more.. so we are going to check if we might have semi valid
  // connection by sorting the probabilities array and see if it has
  // a probability above 60
  probability = probabilities.sort(function sort(a, b) {
    return a.probability - b.probability;
  }).pop();

  if (probability && probability.probability >= 60) {
    fn(undefined, probability.connection);
    return this;
  }

  // well, that didn't work out, so assume failure
  fn(new Error('The connection pool is full'));
  return this;
};

/**
 * Check if a connection is available for writing.
 *
 * @param {net.Connection} net The connection.
 * @param {Boolean} ignore Ignore closed or dead connections.
 * @returns {Number} Probability that his connection is available or will be.
 * @api private
 */
Manager.prototype.isAvailable = function isAvailable(net, ignore) {
  var readyState = net.readyState
    , writable = readyState === 'open' || readyState === 'writeOnly'
    , writePending = net._pendingWriteReqs || 0
    , writeQueue = net._writeQueue || []
    , writes = writeQueue.length || writePending;

  // If the stream is writable and we don't have anything pending we are 100%
  // sure that this stream is available for writing.
  if (writable && writes === 0) return 100;

  // The connection is already closed or has been destroyed, why on earth are we
  // getting it then, remove it from the pool and return 0.
  if (readyState === 'closed' || net.destroyed) {
    this.remove(net);
    return 0;
  }

  // If the stream isn't writable we aren't that sure..
  if (!writable) return 0;

  // The connection is still opening, so we can write to it in the future.
  if (readyState === 'opening') return 70;

  // We have some writes, so we are going to substract that amount from our 100.
  if (writes < 100) return 100 - writes;

  // We didn't find any reliable states of the stream, so we are going to
  // assume something random, because we have no clue, so generate a random
  // number between 0 - 70.
  return Math.floor(Math.random() * 70);
};

/**
 * Release the connection from the connection pool.
 *
 * @param {Stream} net The connection.
 * @param {Boolean} hard Switch between destroySoon or destroy.
 * @returns {Boolean} Was the removal successful
 * @api private
 */
Manager.prototype.release = function release(net, hard) {
  var index = this.pool.indexOf(net);
  if (index === -1) return false;

  // Check if the stream is still open.
  if (net) {
    if (!hard) net.destroySoon();
    else net.destroy();

    // Remove it from the pool.
    this.pool.splice(index, 1);

    // Increase the releases metric.
    this.metrics.releases++;
  }

  return true;
};

// Alias remove to release.
Manager.prototype.remove = Manager.prototype.release;

/**
 * Free dead connections from the pool.
 *
 * @param {Number} keep the amount of connection to keep open
 * @param {Boolean} hard destroy all connections instead of destroySoon
 * @api public
 */
Manager.prototype.free = function free(keep, hard) {
  // Default to 0 if no arguments are supplied.
  keep = +keep || 0;

  // Create a back-up of the pool as we will be removing items from the array
  // and this could cause memory / socket leaks as we are unable to close some
  // connections in the array as the index has moved.
  var pool = this.pool.slice(0)
    , saved = 0;

  for (var i = 0, length = pool.length; i < length; i++) {
    var connection = pool[i]
      , probability = this.isAvailable(connection);

    // This is still a healthy connection, so try we probably just want to keep it.
    if (keep && saved < keep && probability === 100) {
      saved++;
      continue;
    }

    this.release(connection, hard);
  }

  // Clear the back-up
  pool.length = 0;

  // See how much connections are still available.
  return this.emit('free', saved, this.pool.length);
};

/**
 * Iterate over the different connection.
 *
 * @param {Function} callback
 * @param {Mixed} context
 * @api public
 */
Manager.prototype.forEach = function forEach(callback, context) {
  this.pool.forEach(callback, context);
  return this;
};

/**
 * Close the connection pool.
 *
 * @param {Boolean} hard destroy all connections
 * @api public
 */
Manager.prototype.end = function end(hard) {
  this.free(0, hard);

  return this.emit('end');
};

module.exports = Manager;
